src/monattack.cpp

```diff
@@ -185,6 +185,54 @@ void mattack::shriek(monster *z, int index)
     sounds::sound(z->pos(), 50, _("a terrible shriek!"));
 }
 
+void mattack::shriek_alert(monster *z, int index)
+{
+    if( !z->can_act() || z->has_effect("shrieking")) {
+        return;
+    }
+
+    Creature *target = z->attack_target();
+
+    if(g->u.sees( *z )){
+    add_msg( _("The %s begins shrieking!"), z->name().c_str());
+    }
+
+    int dist;
+    if( target == nullptr || (dist = rl_dist( z->pos(), target->pos() )) > 15 ||
+        !z->sees( *target ) ) {
+        return;
+    }
+
+    z->moves -= 150;
+    z->reset_special(index); // Reset timer
+    sounds::sound(z->pos(), 120, _("a piercing wail!"));
+    z->add_effect("shrieking", 10);
+
+}
+
+void mattack::shriek_stun(monster *z, int index)
+{
+    if( !z->can_act() || !z->has_effect("shrieking")) {
+        return;
+    }
+
+    Creature *target = z->attack_target();
+    int dist;
+    if( target == nullptr || (dist = rl_dist( z->pos(), target->pos() )) > 15 ||
+        !z->sees( *target ) ) {
+        return;
+    }
+
+    z->reset_special(index); // Reset timer
+
+    if ( one_in((dist + 2 )) && !(target->is_immune_effect("deaf")) ){
```

Comment:
Something like that.
As long as the area would be rather small (say, 6x6 tiles), you could even brute force the cone by checking each tile for `sees` and angle difference. You can take examples of angle calculations from turret targeting code in creature.cpp.

Brute force calculations are undesirable because of high CPU cost, but NPCs are doing a 13x13 search every turn and the game isn't straining heavily when even 5 of them are following the player.